serverdensity/sd-agent
# SD-2316 Add darwin installer

#### What's this PR do?
Adds a darwin directory with the scripts and configuration files needed to create a standard Mac OS X installer that installs the agent, and create a launchd startup item

#### Where should the reviewer start?
The driving script is `build_installer.sh`. It basically creates a temporary directory where to create the virtualenv and the agent itself, and creates a package with them.

#### How should this be manually tested?
Check out this branch. `cd` into the `darwin` directory and run:

    ./build_installer.sh

After a while, 2 packages should appear in the directory. One of them is a component of the other, in the future I expect to add a secondary component for building a preference pane to start/stop the agent.

You may want to download Pacifist from http://www.charlessoft.com to have a look at the package first. You will see the contents of the temporary directoy, plus extra resources (such as the `preinstall` and `postinstall` scripts).

Run the installer named `Server Density Agent Installer 2.0.4.pkg`. You may hit ⌘L to view the installer logs.

After installing successfully, you will have the agent in `/usr/local/sd-agent` and the startup item `/Library/LaunchDaemons/com.serverdensity.agent.plist`. You will also have a new `sd-agent` user.
You need to launch it with `sudo` as it is intended to be run as non-privileged user `sd-agent`.
Also note that you need to provide `config.cfg`.

#### Any background context you want to provide?
* Man pages for `launchd`, `launchctl`, `launchd.plist`, `pkgbuild`, `productbuild`, `dscl`
* [Daemons and Services Programming Guide](https://developer.apple.com/library/mac/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/Introduction.html#//apple_ref/doc/uid/10000172i-SW1-SW1)
* [OS X Creating packages from the command line](http://thegreyblog.blogspot.com.es/2014/06/os-x-creating-packages-from-command_2.html) Parts I and II
* [Software Delivery Legacy Guide](https://developer.apple.com/legacy/library/documentation/DeveloperTools/Conceptual/SoftwareDistribution4/Introduction/Introduction.html#//apple_ref/doc/uid/TP40004615-CH1-DontLinkElementID_70)
* [Distributing Apps Outside the Mac App Store](https://developer.apple.com/library/mac/documentation/IDEs/Conceptual/AppDistributionGuide/DistributingApplicationsOutside/DistributingApplicationsOutside.html)
* Tagging a commit in master should publish the installer image to github releases, but I haven't been able to verify this from a PR branch. So this steps will need to be verified then. However, if that doesn't work correctly we can always add the deploy step that pushes the image to S3 that I used in the `signing-tests` branch.

#### What are the relevant tickets?
https://serverdensity.atlassian.net/browse/SD-2316
https://serverdensity.atlassian.net/browse/SD-2751


#### Screenshots 
Curiously, the meta package does not provide the name of the package you are installing (as opposed to the component package). So, the following screenshot is kind of generic looking:
![captura de pantalla 2016-05-22 a les 22 36 34](https://cloud.githubusercontent.com/assets/7311/15456349/c90290ba-206d-11e6-92dc-1777e398e381.png)

#### Who should review or be notified about this?
@serverdensity/backend-engineering
